### PR TITLE
Enable SSL of LibCurl linked to C++ library

### PR DESCRIPTION
**Pull request by massakam**
### Motivation

Currently, the C++ client library included in the distributed RPM package can not perform https lookup.

> 2018-11-20 21:17:36.708 ERROR HTTPLookupService:200 | Unable to load SSL engine as default, for url https://pulsar.xxx.xxx.yahoo.co.jp:8443/admin/persistent/prop/global/ns/dest/partitions

In addition, access to the ZTS fails if using Athenz authentication.

> 2018-11-20 21:15:59.937 ERROR ZTSClient:318 | Response failed for url https://zts.xxx.yahoo.co.jp:8443/zts/v1/domain/xxx.xxx.xxx.xxx/token. Error Code 1

I think these are because SSL disabled libcurl is linked to the pulsar library.

### Modifications

- Upgraded libcurl included in the docker images to 7.58.0. This is because we can not build curl using static openssl library in versions lower than 7.58.0.
https://github.com/curl/curl/pull/2234
- Build curl with `--with-ssl` option.
- Changed the order of the libraries to be linked so that curl is ahead of openssl.
https://github.com/curl/curl/issues/2559

### Result

HTTPS lookup and access to the ZTS will work properly.

### CODE REVIEW

**Comment by maskit** [comment]
While you are at it, can you update OpenSSL versions?

File: pulsar-client-cpp/docker/Dockerfile [hunk 1]
```diff
@@ -46,12 +46,12 @@ RUN curl -O -L https://zlib.net/zlib-1.2.11.tar.gz && \
     rm -rf /zlib-1.2.11.tar.gz /zlib-1.2.11
 
 # Compile OpenSSL
-RUN curl -O -L https://github.com/openssl/openssl/archive/OpenSSL_1_0_2l.tar.gz && \
-    tar xvfz OpenSSL_1_0_2l.tar.gz && \
-    cd openssl-OpenSSL_1_0_2l/ && \
+RUN curl -O -L https://github.com/openssl/openssl/archive/OpenSSL_1_0_2q.tar.gz && \
+    tar xvfz OpenSSL_1_0_2q.tar.gz && \
```
Comment (merlimat) [inline pulsar-client-cpp/docker/Dockerfile]: Can this be `1_1_0j` as well or there was a reason for it?

Comment (merlimat) [reply pulsar-client-cpp/docker/Dockerfile]: Thanks!

**Comment by massakam** [comment]
@merlimat I'll try it. My user id is `massakam`.

**Comment by massakam** [comment]
Are the images that need to be pushed these?
```
apachepulsar/pulsar-build:manylinux-cp27-cp27mu
apachepulsar/pulsar-build:manylinux-cp27-cp27m
apachepulsar/pulsar-build:manylinux-cp33-cp33m
apachepulsar/pulsar-build:manylinux-cp34-cp34m
apachepulsar/pulsar-build:manylinux-cp35-cp35m
apachepulsar/pulsar-build:manylinux-cp36-cp36m
apachepulsar/pulsar-build:centos-7
apachepulsar/pulsar-build:debian-9
```

**Comment by merlimat** [comment]
That looks correct and you should have access now.

**Comment by massakam** [comment]
Image `apachepulsar/pulsar-build:manylinux-cp33-cp33m` could not be built due to an error:
```
ln: creating symbolic link `/opt/python/cp33-cp33m/include/python3.3' to `/opt/python/cp33-cp33m/include/python3.3m': No such file or directory
The command '/bin/sh -c ln -s /opt/python/${PYTHON_SPEC}/include/python${PYTHON_VERSION}m /opt/python/${PYTHON_SPEC}/include/python${PYTHON_VERSION}' returned a non-zero code: 1
```
In addition, SSL was not enabled for debian images. I will fix these issues with https://github.com/apache/pulsar/pull/3035.

**Pull request merged**